### PR TITLE
feature: major type updates for ViewModel

### PR DESCRIPTION
**Pull request by davecoates**
- VIewModelFactory make it easier to type partial records. Provide utility types for values & partials
- ViewModelCache update to use new types & more strongly/correctly type caching functions

### CODE REVIEW

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts [hunk 1]
```diff
@@ -479,23 +516,45 @@ export default class ViewModelCache<T extends ViewModelInterface<any, any>> {
      * in which case each entry in the array will be converted to the view model if required
      * and returned.
      */
-    add(
+    add<T extends InstanceType<ViewModelClassType>>(recordOrData: T): T;
+    add<T extends InstanceType<ViewModelClassType>>(recordOrData: T[]): T[];
+    add<T extends InstanceType<ViewModelClassType>>(recordOrData: T): T;
```
Comment (fanglong-alliance) [inline js-packages/@prestojs/viewmodel/src/ViewModelCache.ts]: duplication of 519?

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts [hunk 2]
```diff
@@ -508,6 +567,8 @@ export default class ViewModelCache<T extends ViewModelInterface<any, any>> {
             recordCache = new RecordCache(this.viewModel, this.onAnyChange.bind(this));
             this.cache.set(pkKey, recordCache);
         }
+        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+        // @ts-ignore
         recordCache.add(record);
```
Comment (fanglong-alliance) [inline js-packages/@prestojs/viewmodel/src/ViewModelCache.ts]: what was the complaint here?

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts [hunk 3]
```diff
@@ -622,19 +708,26 @@ export default class ViewModelCache<T extends ViewModelInterface<any, any>> {
...
+        pks: PrimaryKey[],
+        fieldNames: readonly FieldNames[],
+        removeNulls?: boolean
+    ): (PartialViewModel<ViewModelClassType, FieldNames> | null)[];
+    getList<
+        FieldNames extends keyof ViewModelClassType['fields'],
+        T extends InstanceType<ViewModelClassType>
+    >(
+        pksOrRecords: (T | PrimaryKey)[],
+        fieldNames?: readonly FieldNames[] | boolean,
```
Comment (fanglong-alliance) [inline js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:718]: fieldNames being bool isnt mentioned in the docstring?

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts [hunk 4]
```diff
@@ -691,7 +795,7 @@ export default class ViewModelCache<T extends ViewModelInterface<any, any>> {
      *
      * @returns true if anything was removed, false otherwise
      */
-    delete(pk: PrimaryKey, fieldNames?: string[]): boolean {
+    delete(pk: PrimaryKey, fieldNames?: (keyof ViewModelClassType['fields'])[]): boolean {
```
Comment (fanglong-alliance) [inline js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:798]: why not FieldNames?

Comment (fanglong-alliance) [inline js-packages/@prestojs/viewmodel/src/ViewModelCache.ts]: readonly?

File: js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts [hunk 5]
```diff
@@ -157,13 +152,42 @@ export interface ViewModelConstructor<
...
+    // const C: ViewModelConstructor<any> = B;
+    // const a3 = new C({ name: 'test' });
+    // Error: Property 'age' does not exist on type ...
+    // a3.age
+    new (data: FieldDataMappingRaw<FieldMappingType>): ViewModelInterface<
+        FieldMappingType,
+        keyof FieldMappingType,
+        PkFieldType,
+        PkType
+    >;
```
Comment (fanglong-alliance) [inline js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:190]: is this what we discussed here? https://alliance-software.slack.com/archives/CGPEMBTJL/p1587974607004400 :p

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts (hunk 2 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/ViewModelCache.ts]: oops none anymore, i had an original version done differently

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts (hunk 3 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:718]: fieldNames is required to be an array but only if you provide primary keys, not records. this signature here is the combined signature of the overloads so has to accomodate everything and hence when `fieldNames` is `boolean` it's actually the `removeNulls` option. Probably clearer looking at then documentation: https://prestojs.com/docs/viewmodel/ViewModelCache/#method-getList

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts (hunk 1 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/ViewModelCache.ts]: ta

File: js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts (hunk 5 above)
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:190]: yeah basically that

File: js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts [hunk 6]
```diff
@@ -960,3 +984,46 @@ export default function viewModelFactory<T extends FieldsMapping>(
...
+ * Type to describe values of an instance of ViewModel
+ *
+ * Usage:
+ *
+ * ```ts
+ * const Person = viewModelFactory({
+ *   name: new Field<string>(),
+ *   age: new Field<number>(),
+ * });
+ * type PersonValues = RecordValues<typeof Person>;
```
Comment (fanglong-alliance) [inline js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts]: do you mean ViewModelValues?

Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts]: yep ta

File: js-packages/@prestojs/viewmodel/src/ViewModelCache.ts (hunk 3 above)
Comment (fanglong-alliance) [reply js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:718]: ah i see

**Comment by fanglong-alliance** [comment]
looks good!

**Pull request merged**